osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -18,146 +23,285 @@
-                throw new IllegalStateException("Could not create encrypted secret: secret has no key reference");
-            }
-            final byte[] secretBytes = HexUtils.fromHexString(secret.getEncodedSecret());
-            final EncryptedSecret encryptedSecret = new EncryptedSecret(keyReference.getEncryptionProviderType(),
-                    secretBytes);
-            return this.createTypedSecret(secret, keyReference, encryptedSecret);
-        } else {    //Should never happen because of stream mapping in retrieveSecrets()
-            throw new IllegalStateException("Could not create typed secret for NULL secret");
+    public synchronized void storeSecrets(final String deviceIdentification, final List<TypedSecret> secrets) {
+        List<EncryptedTypedSecret> aesSecrets =
+                secrets.stream()
+                       .map(ts -> new EncryptedTypedSecret(ts.getSecret(), ts.getSecretType()))
+                       .map(this::reencryptRsa2Aes)
+                       .collect(toList());
+        this.storeAesSecrets(deviceIdentification, aesSecrets);
+    }
+
+    private void storeAesSecrets(final String deviceIdentification, final List<EncryptedTypedSecret> secrets) {
+        secrets.stream()
+               .map(ets -> this.validateNewSecret(deviceIdentification, ets))
+               .map(ets -> this.createDbEncrypted(deviceIdentification, ets,
+                       this.getKeyByReference(ets.encryptionKeyReference)))
+               .collect(collectingAndThen(toList(), this.secretRepository::saveAll));
+    }
+
+    public synchronized void activateNewSecrets(final String deviceIdentification, final List<SecretType> secretTypes) {
+        secretTypes.stream().forEach(t -> this.checkNrNewSecretsOfType(deviceIdentification, t, 1));
+        secretTypes.stream()
+                   .map(t -> this.getUpdatedSecretsForActivation(deviceIdentification, t))
+                   .flatMap(Collection::stream)
+                   .collect(collectingAndThen(toList(), this.secretRepository::saveAll));
+    }
+
+    public boolean hasNewSecret(final String deviceIdentification, final SecretType secretType) {
+        return this.secretRepository.getSecretCount(deviceIdentification, secretType, SecretStatus.NEW) > 0;
+    }
+
+    private List<DbEncryptedSecret> getUpdatedSecretsForActivation(final String deviceIdentification,
+            final SecretType secretType) {
+        List<DbEncryptedSecret> updatedSecrets = new ArrayList<>();
+        final Optional<DbEncryptedSecret> activeSecretOptional = this
+                .getSingleDbEncryptedSecret(deviceIdentification, secretType, SecretStatus.ACTIVE);
+        if (activeSecretOptional.isPresent()) {
+            final DbEncryptedSecret currentSecret = activeSecretOptional.get();
+            currentSecret.setSecretStatus(SecretStatus.EXPIRED);
+            updatedSecrets.add(currentSecret);
         }
+        final DbEncryptedSecret newSecret = this
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Call "Optional#isPresent()" before accessing the value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS3655)